pkg/controller/nodekeeper/nodekeeper_controller.go

```diff
@@ -0,0 +1,225 @@
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileNodeKeeper{
+		client:               mgr.GetClient(),
+		scheme:               mgr.GetScheme(),
+		metricsClientBuilder: metrics.NewBuilder(),
+	}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("nodekeeper-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource UpgradeConfig, status change will not trigger a reconcile
+	err = c.Watch(&source.Kind{Type: &corev1.Node{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileNodeKeeper implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileNodeKeeper{}
+
+// ReconcileNodeKeeper reconciles a NodeKeeper object
+type ReconcileNodeKeeper struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client               client.Client
+	scheme               *runtime.Scheme
+	metricsClientBuilder metrics.MetricsBuilder
+}
+
+// Reconcile reads that state of the cluster for a UpgradeConfig object and makes changes based on the state read
+// and what is in the UpgradeConfig.Spec
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileNodeKeeper) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling NodeKeeper")
+
+	// Determine if the cluster is upgrading
+	yes, err := isClusterUpgrading(r.client, reqLogger)
+	if err != nil {
+		// Return the error that occurred.
+		return reconcile.Result{}, err
+	} else if !yes {
+		// Cluster is not upgrading, no error occurred.
+		return reconcile.Result{}, nil
+	}
+
+	reqLogger.Info("Cluster is upgrading. Proceeding.")
+
+	// Initialise metrics
+	metricsClient, err := r.metricsClientBuilder.NewClient(r.client)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	pdbForceDrainTimeout, err = getPDBForceDrainTimeout(r.client, reqLogger)
+	if err != nil {
+		if err == upgradeConfigNotfound {
+			reqLogger.Info(upgradeConfigNotfound.Error())
+			reqLogger.Info("Requeuing")
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	reqLogger.Info("Checking for PodDisruptionBudget alerts.")
+
+	found, err := checkPDBAlerts(metricsClient)
```

Comment:
Can we bring the current implmentation into here?
https://github.com/openshift/managed-upgrade-operator/blob/ac2c24dcb55cd90ce8dab2e7395b0b27e81b0d3e/pkg/osd_cluster_upgrader/upgrader.go#L242-L251

https://github.com/openshift/managed-upgrade-operator/blob/ac2c24dcb55cd90ce8dab2e7395b0b27e81b0d3e/pkg/osd_cluster_upgrader/upgrader.go#L278

Reply:
Retro fit for this is here: https://issues.redhat.com/browse/OSD-4716 will include alert + actual drain.